### PR TITLE
test: add new e2e test cases for kubectl plugin

### PR DESCRIPTION
**Pull request by venkateshjayagopal**
**What this PR does / why we need it**:

Adding new test cases for testing "kubectl runtime-enforcer plugin"

**Which issue(s) this PR fixes**

fixes #N/A

**Special notes for your reviewer**:
In-progress

**Checklist**:

- [x] squashed commits into logical changes
- [x] includes documentation
- [x] adds unit tests
- [x] adds or updates e2e tests

### CODE REVIEW

File: test/e2e/e2e_suite_test.go [hunk 1]
```diff
@@ -257,3 +274,171 @@ func installHelmRepos(charts []helmChart) env.Func {
 		return ctx, nil
 	}
 }
+
+// installKubectlRuntimeEnforcer downloads and installs the kubectl runtime-enforcer plugin.
+func installKubectlRuntimeEnforcer() env.Func {
```
Comment (andreaterzolo) [inline test/e2e/e2e_suite_test.go]: I'm just wondering if it wouldn't be easier to just build the kubectl plugin locally in the makefile and using that binary instead of downloading a fixed release here. Please note that these e2e also run on each PRs, so we probably want to test the latest version of the plugin.

What i would do is to just add this line to the makefile, so that the plugin becomes a dependency of the e2e test

```diff
diff --git a/Makefile b/Makefile
index 76167f9..1a12702 100644
--- a/Makefile
+++ b/Makefile
@@ -125,6 +125,8 @@ PLUGIN_PLATFORMS ?= linux/amd64 linux/arm64 darwin/amd64 darwin/arm64
 kubectl-plugin: ## Build kubectl plugin for the current platform.
        go build -ldflags "-X github.com/rancher-sandbox/runtime-enforcer/internal/kubectlplugin.version=$(KUBECTL_PLUGIN_VERSION)" -o ./bin/kubectl-runtime_enforcer ./cmd/kubectl-plugin
 
+E2E_DEPS += kubectl-plugin
+
 .PHONY: kubectl-plugin-cross
 kubectl-plugin-cross: ## Build kubectl plugin for all target platforms.
        @mkdir -p bin/kubectl-plugin
```

and once built we can directly call it like `./bin/kubectl-runtime_enforcer` without the need of moving it into the PATH using `kubectl runtime-enforcer`. WDYT?

File: test/e2e/kubectl_plugin_test.go
Comment (andreaterzolo) [inline test/e2e/kubectl_plugin_test.go:1]: This is really great! Thank you for all these tests! 
To be honest i would slightly change approach. As you can see some of the cases you tested here are already covered in our unit tests `internal/kubectlplugin`. 

As you can see, we have specific test for each action:
* internal/kubectlplugin/policy_exec_test.go
* internal/kubectlplugin/policy_mode_test.go
* ...

What i would do is maybe keeping a unique smoke test here like 
```go
func TestKubectlPluginSmokeTest(t *testing.T) {
``` 
which tests a couple of commands just to be sure the binary is working (promote + change policy + add binary) and then move each specific case to the corresponding unit test where needed. 
Some stuff should be already be there but maybe you have identified extra corner cases to add there. WDYT? 
@holyspectral  @kyledong-suse any thoughts here? i would avoid to test the same logic in 2 places

**Review by Andreagit97 — COMMENTED** [review]
Thank you for all these tests!
Just left some high level comments on the architecture

File: test/e2e/e2e_suite_test.go (hunk 1 above)
Comment (venkateshjayagopal) [reply test/e2e/e2e_suite_test.go]: Thats a good idea. I will implement the same.

Comment (venkateshjayagopal) [reply test/e2e/e2e_suite_test.go]: @Andreagit97 When I was trying to implement the change I figured out that we use "make test-e2e" to run e2e tests for PRs.
In QA we use "E2E_USE_EXISTING_CLUSTER=true" to make use of the existing cluster to run the test. 
To bridge the gap, I am planning to implement performing "make kubectl-plugin" when using "E2E_USE_EXISTING_CLUSTER=true".
OR should I make "make kubectl-plugin" as default both the scenarios?.

Comment (andreaterzolo) [reply test/e2e/e2e_suite_test.go]: > OR should I make "make kubectl-plugin" as default both the scenarios?

This seems reasonable i don't see issues in having it in both scenarios

Comment (venkateshjayagopal) [reply test/e2e/e2e_suite_test.go]: Added makeKubectlPlugin() func which performs "make kubectl-plugin" .
Calling this function at TestMain() function.

File: test/e2e/kubectl_plugin_test.go
Comment (venkateshjayagopal) [reply test/e2e/kubectl_plugin_test.go:1]: @Andreagit97 I did see the unit test covers some scenario under internal/kubectlplugin. When running tests in QA repo we are running tests only under `test/e2e/` directory.
I dont think we are running cases redundantly. If you are ok, please resolve this conversation.

Comment (andreaterzolo) [reply test/e2e/kubectl_plugin_test.go:1]: > When running tests in QA repo we are running tests only under test/e2e/ directory.

 yep i understand this but my point is more "should we test every single feature of the plugin in e2e tests? do we expect the behavior of the plugin to change on different cluster setups?"
IMO we should test that the plugin is at least working in the cluster with a smoke test on the main commands, but the core testing logic of all possible cases should reside in the unit tests where we can quickly iterate over fixes.
But of course this is just my opinion, curios to know what other thinks about that

File: test/e2e/e2e_suite_test.go [hunk 2]
```diff
@@ -49,6 +51,19 @@ func installDependencies() bool {
 	return os.Getenv("E2E_SKIP_DEPENDENCIES") != "true"
 }
 
+func makeKubectlPlugin() error {
```
Comment (andreaterzolo) [inline test/e2e/e2e_suite_test.go]: why don't we do this into the makefile?
```diff
diff --git a/Makefile b/Makefile
index 76167f9..7c82a04 100644
--- a/Makefile
+++ b/Makefile
@@ -93,7 +93,7 @@ helm-unittest:
        helm unittest charts/runtime-enforcer/ --file "tests/**/*_test.yaml"
 
 .PHONY: test-e2e
-test-e2e:
+test-e2e: kubectl-plugin
 ifneq ($(E2E_USE_EXISTING_CLUSTER),true)
 ifeq ($(E2E_NO_REBUILD),)
        TAG=latest make $(E2E_DEPS)
```

File: test/e2e/kubectl_plugin_test.go
Comment (venkateshjayagopal) [reply test/e2e/kubectl_plugin_test.go:1]: Removed idempotent, --dry-run and error handling cases for kubectl plugin. Now e2e covers only policy promote, adding and denying executables from the policy.
the run time of test now decreased to almost 2 mins approx.

```
neuvector@Host-01-RKE2:~/runtime-enforcer-vj$ sudo E2E_USE_EXISTING_CLUSTER=true E2E_SKIP_DEPENDENCIES=true /usr/local/go/bin/go test -v -timeout 20m -run TestKubectl ./test/e2e/ -count=1
go build -ldflags "-X github.com/rancher-sandbox/runtime-enforcer/internal/kubectlplugin.version=8307d01-dirty" -o ./bin/kubectl-runtime_enforcer ./cmd/kubectl-plugin
{"time":"2026-06-12T15:59:10.609954265Z","level":"INFO","msg":"uninstall helm release if present","name":"runtime-enforcer","namespace":"run-enf-e2e-runtime-enforcer"}
{"time":"2026-06-12T15:59:10.757847297Z","level":"INFO","msg":"remove helm repo if present","repo":"run-enf-e2e-runtime-enforcer-repo"}
{"time":"2026-06-12T15:59:10.882994517Z","level":"INFO","msg":"installing helm release","path":"../../charts/runtime-enforcer/","name":"runtime-enforcer","namespace":"run-enf-e2e-runtime-enforcer"}
=== RUN   TestKubectlPluginProposalPromoteTest
    e2e_test.go:69: test kubectl runtime-enforcer proposal promote PROPOSAL_NAME [flags]
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote
    helpers_test.go:39: setup shared k8s client
    e2e_test.go:71: creating test namespace: "run-enf-e2e-286fbb60373ebd4bde10"
    kubectl_plugin_test.go:32: installing test Opensuse deployment
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/required_resources_become_available
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/the_workload_proposal_is_created_successfully_for_the_opensuse_pod
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/the_running_process_is_learned
    kubectl_plugin_test.go:72: waiting for policy proposal to be created:  deploy-opensuse-deployment
    kubectl_plugin_test.go:84: waiting for processes to be learned
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/kubectl_plugin_promotes_proposal_successfully
    kubectl_plugin_test.go:115: stdout: Promoted WorkloadPolicyProposal "deploy-opensuse-deployment" in namespace "run-enf-e2e-286fbb60373ebd4bde10" to WorkloadPolicy.
        WorkloadPolicy "deploy-opensuse-deployment" in namespace "run-enf-e2e-286fbb60373ebd4bde10" has been created.
    kubectl_plugin_test.go:116: stderr:
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/WorkloadPolicy_was_created_in_monitor_mode
=== RUN   TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/WorkloadPolicyProposal_was_deleted_after_promotion
=== NAME  TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote
    kubectl_plugin_test.go:166: deleting test Opensuse deployment
    kubectl_plugin_test.go:166: waiting for Opensuse deployment to be deleted
--- PASS: TestKubectlPluginProposalPromoteTest (40.88s)
    --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote (40.88s)
        --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/required_resources_become_available (10.01s)
        --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/the_workload_proposal_is_created_successfully_for_the_opensuse_pod (5.01s)
        --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/the_running_process_is_learned (5.01s)
        --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/kubectl_plugin_promotes_proposal_successfully (0.64s)
        --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/WorkloadPolicy_was_created_in_monitor_mode (5.01s)
        --- PASS: TestKubectlPluginProposalPromoteTest/kubectl_plugin:_proposal_promote/WorkloadPolicyProposal_was_deleted_after_promotion (5.01s)
=== RUN   TestKubectlPluginPolicyModeTest
    e2e_test.go:75: test kubectl runtime-enforcer policy monitor POLICY_NAME [flags]
=== RUN   TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode
    helpers_test.go:39: setup shared k8s client
    e2e_test.go:77: creating test namespace: "run-enf-e2e-4800504e599b064a3eca"
    kubectl_plugin_test.go:203: creating workload policy "run-enf-e2e-4800504e599b064a3eca/mode-test-policy" and waiting for it to become Ready
    helpers_test.go:135: checking workloadpolicy status: {1 map[] 3 3 0 0 [] Ready 0 []}
=== RUN   TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode/required_resources_become_available
=== RUN   TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode/kubectl_plugin_switches_mode_to_protect
    kubectl_plugin_test.go:219: stdout: Successfully set WorkloadPolicy "mode-test-policy" in namespace "run-enf-e2e-4800504e599b064a3eca" to "protect" mode.
    kubectl_plugin_test.go:220: stderr:
=== RUN   TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode/kubectl_plugin_switches_mode_back_to_monitor
    kubectl_plugin_test.go:252: stdout: Successfully set WorkloadPolicy "mode-test-policy" in namespace "run-enf-e2e-4800504e599b064a3eca" to "monitor" mode.
    kubectl_plugin_test.go:253: stderr:
=== NAME  TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode
    kubectl_plugin_test.go:278: deleting workload policy "run-enf-e2e-4800504e599b064a3eca/mode-test-policy" and waiting for it to be deleted
--- PASS: TestKubectlPluginPolicyModeTest (20.35s)
    --- PASS: TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode (20.35s)
        --- PASS: TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode/required_resources_become_available (10.01s)
        --- PASS: TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode/kubectl_plugin_switches_mode_to_protect (0.14s)
        --- PASS: TestKubectlPluginPolicyModeTest/kubectl_plugin:_policy_mode/kubectl_plugin_switches_mode_back_to_monitor (0.14s)
=== RUN   TestKubectlPluginPolicyExecAllowTest
    e2e_test.go:81: test kubectl runtime-enforcer policy allow POLICY_NAME <container-name> <executable-name> [<executable-name>...] [flags]
=== RUN   TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow
    helpers_test.go:39: setup shared k8s client
    e2e_test.go:85: creating test namespace: "run-enf-e2e-74ac059c83372c314f5e"
    kubectl_plugin_test.go:311: creating workload policy "run-enf-e2e-74ac059c83372c314f5e/exec-allow-test-policy" and waiting for it to become Ready
    helpers_test.go:135: checking workloadpolicy status: {1 map[] 3 3 0 0 [] Ready 0 []}
=== RUN   TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow/required_resources_become_available
=== RUN   TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow/kubectl_plugin_allows_new_executables
    kubectl_plugin_test.go:327: stdout: Successfully updated executables for WorkloadPolicy "exec-allow-test-policy" in namespace "run-enf-e2e-74ac059c83372c314f5e".
    kubectl_plugin_test.go:328: stderr:
=== NAME  TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow
    kubectl_plugin_test.go:359: deleting workload policy "run-enf-e2e-74ac059c83372c314f5e/exec-allow-test-policy" and waiting for it to be deleted
--- PASS: TestKubectlPluginPolicyExecAllowTest (20.22s)
    --- PASS: TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow (20.22s)
        --- PASS: TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow/required_resources_become_available (10.02s)
        --- PASS: TestKubectlPluginPolicyExecAllowTest/kubectl_plugin:_policy_exec_allow/kubectl_plugin_allows_new_executables (0.14s)
=== RUN   TestKubectlPluginPolicyExecDenyTest
    e2e_test.go:89: test kubectl runtime-enforcer policy deny POLICY_NAME <container-name> <executable-name> [<executable-name>...] [flags]
=== RUN   TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny
    helpers_test.go:39: setup shared k8s client
    e2e_test.go:93: creating test namespace: "run-enf-e2e-f95935906f8911297d27"
    kubectl_plugin_test.go:395: creating workload policy "run-enf-e2e-f95935906f8911297d27/exec-deny-test-policy" and waiting for it to become Ready
    helpers_test.go:135: checking workloadpolicy status: {1 map[] 3 3 0 0 [] Ready 0 []}
=== RUN   TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny/required_resources_become_available
=== RUN   TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny/kubectl_plugin_denies_executables
    kubectl_plugin_test.go:411: stdout: Successfully updated executables for WorkloadPolicy "exec-deny-test-policy" in namespace "run-enf-e2e-f95935906f8911297d27".
    kubectl_plugin_test.go:412: stderr:
=== NAME  TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny
    kubectl_plugin_test.go:444: deleting workload policy "run-enf-e2e-f95935906f8911297d27/exec-deny-test-policy" and waiting for it to be deleted
--- PASS: TestKubectlPluginPolicyExecDenyTest (20.22s)
    --- PASS: TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny (20.22s)
        --- PASS: TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny/required_resources_become_available (10.02s)
        --- PASS: TestKubectlPluginPolicyExecDenyTest/kubectl_plugin:_policy_exec_deny/kubectl_plugin_denies_executables (0.14s)
PASS
{"time":"2026-06-12T16:01:21.538083302Z","level":"INFO","msg":"uninstall helm release if present","name":"runtime-enforcer","namespace":"run-enf-e2e-runtime-enforcer"}
{"time":"2026-06-12T16:01:22.491367727Z","level":"INFO","msg":"remove helm repo if present","repo":"run-enf-e2e-runtime-enforcer-repo"}
ok      github.com/rancher-sandbox/runtime-enforcer/test/e2e    133.202s
```

File: test/e2e/kubectl_plugin_test.go [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+				err := wait.For(conditions.New(r).ResourceMatch(&policy, func(_ k8s.Object) bool {
+					return policy.Spec.Mode == policymode.MonitorString
+				}), wait.WithTimeout(10*time.Second))
+				require.NoError(t, err)
+
+				assert.Equal(t, policymode.MonitorString, policy.Spec.Mode)
+				assert.NotNil(t, policy.Spec.RulesByContainer["opensuse"])
+
+				return ctx
+			}).
```
Comment (andreaterzolo) [inline test/e2e/kubectl_plugin_test.go:146]: to reduce execution time i'm wondering if we cannot just append the other cases here. Instead of having 4 different e2e tests, we can just have only one smoke e2e test where we assert the full flow. 

So now in this test you are testing the promotion. 
What about adding the other assertions after the current promotion?
* we promote the policy
* we try to switch the mode to protect and viceversa
* we add a binary
* we remove a binary
* we cleanup everything

WDYT?

**Review by Andreagit97 — COMMENTED** [review]
Thanks, just a couple of comments

File: test/e2e/e2e_suite_test.go (hunk 2 above)
Comment (venkateshjayagopal) [reply test/e2e/e2e_suite_test.go]: @Andreagit97 We dont execute E2E tests with "make test-e2e". I explained in this thread  https://github.com/rancher-sandbox/runtime-enforcer/pull/671#discussion_r3391443304

File: test/e2e/kubectl_plugin_test.go (hunk 3 above)
Comment (venkateshjayagopal) [reply test/e2e/kubectl_plugin_test.go:146]: Yes I moved all asserts under one test called TestKubectlPluginSmokeTest 

```
--- PASS: TestKubectlPluginSmokeTest (41.43s)
    --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote (41.43s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/required_resources_become_available (10.02s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/the_workload_proposal_is_created_successfully_for_the_opensuse_pod (5.01s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/the_running_process_is_learned (5.01s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/kubectl_plugin_promotes_proposal_successfully (0.64s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/WorkloadPolicy_was_created_in_monitor_mode (5.01s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/WorkloadPolicyProposal_was_deleted_after_promotion (5.00s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/kubectl_plugin_switches_mode_to_protect (0.14s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/kubectl_plugin_switches_mode_back_to_monitor (0.14s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/kubectl_plugin_allows_new_executables (0.14s)
        --- PASS: TestKubectlPluginSmokeTest/kubectl_plugin:_proposal_promote/kubectl_plugin_denies_executables (0.14s)
PASS
{"time":"2026-06-16T17:07:01.115753597Z","level":"INFO","msg":"uninstall helm release if present","name":"runtime-enforcer","namespace":"run-enf-e2e-runtime-enforcer"}
{"time":"2026-06-16T17:07:01.996048042Z","level":"INFO","msg":"remove helm repo if present","repo":"run-enf-e2e-runtime-enforcer-repo"}
ok      github.com/rancher-sandbox/runtime-enforcer/test/e2e    72.869s
```

File: test/e2e/e2e_suite_test.go (hunk 2 above)
Comment (andreaterzolo) [reply test/e2e/e2e_suite_test.go]: oh i see, can i ask you why? 
instead of calling https://github.com/venkateshjayagopal/runtime-enforcer-e2e/blob/ced6fbdcee1437c62d6f4a002afcd9a6d735b6bf/.github/workflows/e2e.yml#L47
```bash
E2E_USE_EXISTING_CLUSTER=true E2E_SKIP_DEPENDENCIES= go test -v -timeout 20m ./test/e2e/
```

you could do

```bash
make test-e2e E2E_USE_EXISTING_CLUSTER=true E2E_SKIP_DEPENDENCIES=
```

Minor note: `E2E_SKIP_DEPENDENCIES=` has no effect, if you really want to skip dependencies you should use `E2E_SKIP_DEPENDENCIES=true`

Comment (venkateshjayagopal) [reply test/e2e/e2e_suite_test.go]: @Andreagit97 Due to current E2E ticket https://github.com/rancher-sandbox/runtime-enforcer/issues/273 is blocked due to infra challenge, we might end up running E2E tests on the cluster which is up and running (i.e running the test on self hosted github runner). I was trying to understand if its necessary to execute "make" command to compile and create binary which is not needed (bpf, debugger etc). We just need the "kubectl_runtime-enforcer" binary for these tests.

Thanks for the note, we need the dependencies to be installed. I will set it as "false".

**Review by Andreagit97 — APPROVED** [review]
Thank you for addressing the comments! Great job!
